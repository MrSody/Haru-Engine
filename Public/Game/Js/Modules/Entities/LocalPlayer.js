class LocalPlayer extends Player {
    constructor (datos) {
        super(datos);
        this.money = datos.money;

        this.absPos = {absX: 0, absY: 0};
        this.goRun = false;
        this.path = [[]];
    }

/* ------------------------------ *
    GETTERS
* ------------------------------ */
    getAbsPos () {
        return this.absPos;
    }

    isRunning () {
        return this.goRun;
    }

/* ------------------------------ *
    SETTERS
* ------------------------------ */
    setAbsPos (absX, absY) {
        this.absPos.absX = absX;
        this.absPos.absY = absY;
    }

    setRun (running) {
        this.goRun = running;
    }

    setPath (path) {
		this.path = path;
        this.moving = true;
        
        if (this.goRun) {
            this.mode = 2;
        } else {
            this.mode = 1;
        }

        // Check if on the way to attack
		if ((this.goAttack || this.goToNpc) && this.stepCount == 0) {
			// Face player towards enemy
			if (this.path[this.path.length - 1][0] > this.path[this.path.length - 2][0]) {
				this.finalDir = 1;
			} else if (this.path[this.path.length - 1][0] < this.path[this.path.length - 2][0]) {
				this.finalDir = 3;
			} else if (this.path[this.path.length - 1][1] > this.path[this.path.length - 2][1]) {
				this.finalDir = 2;
			} else if (this.path[this.path.length - 1][1] < this.path[this.path.length - 2][1]) {
				this.finalDir = 0;
			}

			// Remove last path element so player doesn't step on enemy
			this.path.pop();
		}
    }

/* ------------------------------ *
    FUNCIONES
* ------------------------------ */
    playerWalking () {
        // Path length is 1 i.e. when clicked on player position
        if (this.path.length >= 1) {
            if (this.stepCount != 0) {

                let posX = this.path[this.stepCount][0],
                    posY = this.path[this.stepCount][1],
                    lastPosX = this.path[this.stepCount - 1][0],
                    lastPosY = this.path[this.stepCount - 1][1];

                if (posX < lastPosX) { // Left
                    this.dir = 3;
                    this.absPos.absX = -1;
                    this.absPos.absY = 0;

                } else if(posX > lastPosX) { // Right
                    this.dir = 1;
                    this.absPos.absX = 1;
                    this.absPos.absY = 0;

                } else if(posY < lastPosY) { // Up
                    this.dir = 0;
                    this.absPos.absX = 0;
                    this.absPos.absY = -1;

                } else if(posY > lastPosY) { // Down
                    this.dir = 2;
                    this.absPos.absX = 0;
                    this.absPos.absY = 1;

                }
            } else {
                this.absPos.absX = 0;
                this.absPos.absY = 0;
            }

            if (this.stepCount < this.path.length - 1 && !this.moveInterrupt) {
                this.stepCount++;
            } else { // End of path
                if(this.goFight != null) {
                    this.fighting = this.goFight;
                    this.mode = 3;
                } else if(this.goToNpc != null) {
                    this.talkingTo = this.goToNpc;
                }
                this.moving = false;
                this.mode = 0;
                this.path.splice(0, this.stepCount);
                //this.stepSnd.pause();
                this.stepCount = 0;
                //this.dir = this.finalDir;
            }
        }
    }

    playerRunning () {
        // Path length is 1 i.e. when clicked on player position
        if (this.path.length >= 1) {
            if (this.stepCount != 0) {

                let lastCount;

                if (this.stepCount - 1 == 0) {
                    lastCount = this.stepCount - 1;
                } else {
                    lastCount = this.stepCount - 2;
                }

                let posX = this.path[this.stepCount][0],
                    posY = this.path[this.stepCount][1],
                    lastPosX = this.path[lastCount][0],
                    lastPosY = this.path[lastCount][1];

                if (posX < lastPosX) { // Left
                    if (posY < lastPosY) {
                        this.dir = 3;
                        this.absPos.absX = -1;
                        this.absPos.absY = -1;
                    } else if (posY > lastPosY) {
                        this.dir = 3;
                        this.absPos.absX = -1;
                        this.absPos.absY = 1;
                    } else {
                        this.dir = 3;
                        this.absPos.absX = -2;
                        this.absPos.absY = 0;
                    }

                } else if(posX > lastPosX) { // Right
                    if (posY < lastPosY) {
                        this.dir = 1;
                        this.absPos.absX = 1;
                        this.absPos.absY = -1;
                    } else if (posY > lastPosY) {
                        this.dir = 1;
                        this.absPos.absX = 1;
                        this.absPos.absY = 1;
                    } else {
                        this.dir = 1;
                        this.absPos.absX = 2;
                        this.absPos.absY = 0;
                    }

                } else if(posY < lastPosY) { // Up
                    if (posY < lastPosY) {
                        this.dir = 0;
                        this.absPos.absX = 1;
                        this.absPos.absY = -1;
                    } else if (posY > lastPosY) {
                        this.dir = 0;
                        this.absPos.absX = -1;
                        this.absPos.absY = -1;
                    } else {
                        this.dir = 0;
                        this.absPos.absX = 0;
                        this.absPos.absY = -2;
                    }

                } else if(posY > lastPosY) { // Down
                    if (posY < lastPosY) {
                        this.dir = 2;
                        this.absPos.absX = 1;
                        this.absPos.absY = 1;
                    } else if (posY > lastPosY) {
                        this.dir = 2;
                        this.absPos.absX = -1;
                        this.absPos.absY = 1;
                    } else {
                        this.dir = 2;
                        this.absPos.absX = 0;
                        this.absPos.absY = 2;
                    }
                }
            } else {
                this.absPos.absX = 0;
                this.absPos.absY = 0;
            }

            if (this.stepCount < this.path.length - 1 && !this.moveInterrupt) {
                this.stepCount += 2;
            }
        }
    }

    playerMove () {
        /*
        if(this.stepSnd.paused) {
            this.stepSnd.currentTime = 0;
            this.stepSnd.play();
        }
        */

        // Check if on the way to attack
        this.setPath(this.path);

        switch (this.mode) {
            case 1:
                this.playerWalking();
                break;
            case 2:
                if (((this.path.length - 1) - this.stepCount) > 0) {
                    this.playerRunning();
                } else {
                    if (this.stepCount == this.path.length) {
                        this.path.splice(0, this.stepCount - 2);
                        this.stepCount = 0;
                        this.setRun(false);
                        this.playerWalking();
                    } else {
                        this.setRun(true);
                    }
                }
                break;
        }  
    }

/* ------------------------------ *
    DRAW
* ------------------------------ */

    draw (ctx, HUB, cXnull, cYnull) {
        const tileSize = 32;
        cXnull *= tileSize;
        cYnull *= tileSize;

        this.updateFrame();

        HUB.fillStyle = "#FFF";
        HUB.font = "9pt Minecraftia";
        // Muestra el nombre del player
        HUB.fillText(this.name, cXnull, (cYnull - 20));

        this.drawMode(ctx, cXnull, cYnull);

        let hitDelta = Date.now() - this.lastHitPointUpdate;

        if (this.hitArray.length > 0 && hitDelta > 50) {
            for (var i = 0, j = this.hitArray.length; i < j; i+=1) {
                if (this.hitArray[i] != null) {
                    this.lastHitPointUpdate = Date.now();
                    this.hitArray[i][1] = Math.round((this.hitArray[i][1]-0.1)*10)/10;

                    if (this.hitArray[i][1] <= 0) {
                        this.hitArray.splice(i,1);
                    }
                }
            }
        } else if (this.hitArray.length == 0) {
            this.gotHit = false;
        }

        if (this.hitArray.length > 0) {
            for (var i = 0, max = this.hitArray.length; i < max; i += 1) {
                ctx.fillStyle = 'rgba(255,0,0,'+this.hitArray[i][1]+')';
                ctx.font = "14pt Minecraftia";
                ctx.fillText(this.hitArray[i][0], this.pos["x"]+12, this.pos["y"]-20);
            }
        }
    }
}