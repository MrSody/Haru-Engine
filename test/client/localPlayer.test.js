/**
 * @jest-environment jsdom
 */

const CONSTANT = require('./constants');
const { createCanvas, loadImage } = require('canvas');

import { response } from 'express';
import LocalPlayer from '../../client/public/game/js/modules/entities/player/localPlayer';
import DirectionsEnums from "../../client/public/game/js/modules/enums/directions";

test('setAbsPos', () => {
    let player = new LocalPlayer(CONSTANT.DATA_PLAYER);

    player.setAbsPos(9, 12);

    expect(player.absPos).toStrictEqual({x: 9, y: 12});
});


describe('setPath', () => {

    test('setPath', () => {
        let player = new LocalPlayer(CONSTANT.DATA_PLAYER);
        let responsePath = [[9, 11],
                            [9, 12],
                            [9, 13],
                            [10, 13],
                            [11, 13],
                            [11, 14]];

        player.setPath(CONSTANT.Path);

        expect(player.path).toStrictEqual(responsePath);
    });

    test('setPath - goAttack', () => {
        const player = new LocalPlayer(CONSTANT.DATA_PLAYER);
        let responsePath = [[9, 11],
                            [9, 12],
                            [9, 13],
                            [10, 13],
                            [11, 13]];

        player.goAttack = true;

        player.setPath(CONSTANT.Path);

        expect(player.path).toStrictEqual(responsePath);
        expect(player.finalDirection).toStrictEqual(DirectionsEnums.directions().Down);
    });

    test('setPath - goToNpc', () => {
        const player = new LocalPlayer(CONSTANT.DATA_PLAYER);
        let responsePath = [[9, 11],
                            [9, 12],
                            [9, 13],
                            [10, 13],
                            [11, 13]];

        player.goToNpc = true;

        player.setPath(CONSTANT.Path);

        expect(player.path).toStrictEqual(responsePath);
        expect(player.finalDirection).toStrictEqual(DirectionsEnums.directions().Down);
    });
});

describe('updateAbsPos', () => {
    test('updateAbsPos - Steps - 1', () => {
        const player = new LocalPlayer(CONSTANT.DATA_PLAYER);

        player.setPath(CONSTANT.Path);
        player.stepCount = 1;

        player.updateAbsPos();

        expect(player.dir).toStrictEqual(DirectionsEnums.directions().Down);
        expect(player.absPos).toStrictEqual({x: 0, y: 1});
    });

    test('updateAbsPos - Steps - 2', () => {
        const player = new LocalPlayer(CONSTANT.DATA_PLAYER);

        player.setPath(CONSTANT.Path);
        player.stepCount = 2;

        player.updateAbsPos();

        expect(player.dir).toStrictEqual(DirectionsEnums.directions().Down);
        expect(player.absPos).toStrictEqual({x: 0, y: 1});
    });

    test('updateAbsPos - Steps - 3', () => {
        const player = new LocalPlayer(CONSTANT.DATA_PLAYER);

        player.setPath(CONSTANT.Path);
        player.stepCount = 3;

        player.updateAbsPos();

        expect(player.dir).toStrictEqual(DirectionsEnums.directions().Right);
        expect(player.absPos).toStrictEqual({x: 1, y: 0});
    });

    test('updateAbsPos - Steps - 4', () => {
        const player = new LocalPlayer(CONSTANT.DATA_PLAYER);

        player.setPath(CONSTANT.Path);
        player.stepCount = 4;

        player.updateAbsPos();

        expect(player.dir).toStrictEqual(DirectionsEnums.directions().Right);
        expect(player.absPos).toStrictEqual({x: 1, y: 0});
    });

    test('updateAbsPos - Steps - 5', () => {
        const player = new LocalPlayer(CONSTANT.DATA_PLAYER);

        player.setPath(CONSTANT.Path);
        player.stepCount = 5;

        player.updateAbsPos();

        expect(player.dir).toStrictEqual(DirectionsEnums.directions().Down);
        expect(player.absPos).toStrictEqual({x: 0, y: 1});
    });
});

test('playerWalking', () => {
    const player = new LocalPlayer(CONSTANT.DATA_PLAYER);

    player.setPath(CONSTANT.Path);
    player.stepCount = 1;
    player.tellCount = 1;

    player.playerWalking();

    expect(player.stepCount).toStrictEqual(2);
});

test('playerMove', () => {
    const player = new LocalPlayer(CONSTANT.DATA_PLAYER);

    player.path = CONSTANT.Path;

    player.playerMove(1);
    player.posWorld.x += player.absPos.x;
    player.posWorld.y += player.absPos.y; 
    
    player.playerMove(1);
    player.posWorld.x += player.absPos.x;
    player.posWorld.y += player.absPos.y;
    
    player.playerMove(1);
    player.posWorld.x += player.absPos.x;
    player.posWorld.y += player.absPos.y;
    
    player.playerMove(1);
    player.posWorld.x += player.absPos.x;
    player.posWorld.y += player.absPos.y;

    player.playerMove(1);
    player.posWorld.x += player.absPos.x;
    player.posWorld.y += player.absPos.y;

    player.playerMove(1);
    player.posWorld.x += player.absPos.x;
    player.posWorld.y += player.absPos.y;

    expect(player.posWorld).toStrictEqual({x: 11, y: 14 });
});

test('draw', async () => {
    const player = new LocalPlayer(CONSTANT.DATA_PLAYER);
    const coordinateX = 5;
    const coordinateY = 5;
    const tileSize = 32;

    const canvaCharacters = createCanvas(544, 352);
    const ctxCharacters = canvaCharacters.getContext('2d');
    const canvaHUB = createCanvas(544, 352);
    const ctxHUB = canvaHUB.getContext('2d');

    const image = await loadImage(`${__dirname}\\spritePlayer.png`);

    player.skinBase = image;

    player.draw(ctxCharacters, ctxHUB, coordinateX, coordinateY, tileSize);

    const dataURLCharacters = canvaCharacters.toDataURL();
    expect(dataURLCharacters).toMatchSnapshot();

    const dataURLHUB = canvaHUB.toDataURL();
    expect(dataURLHUB).toMatchSnapshot();
});