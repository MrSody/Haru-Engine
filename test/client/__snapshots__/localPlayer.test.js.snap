// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`draw 1`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAiAAAAFgCAYAAACCOcLtAAAABmJLR0QA/wD/AP+gvaeTAAAIg0lEQVR4nO3dT4ibaR0H8O8kU8m6i6KCSxEEjUoRkSKLzqUwx+zJm44nB0TZY49zzMHDiHvoRZBKdTw19dQ9NZeFgV6yIMuwHnZRRypYdgurFbfVtNtkPMwmTeZP28lkkmfK5wOlybxvnryXKd8+z+/5PQkAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAMC4hXk/AKfbdno7s/y+eqpPvL6d3oyeZFc9Vb9DABNYnPcDwLOqp5rOpbUn33OxOvMQAsDR+d8bxzKrGZDR8HG+0Tjwnq12O0mydHF9ZiHEDAjAZCrzfgA4isPCx9OuAVAWAYTiDWY/niVgnG800rm09tRaEQDmSwABAGZOAAEAZk4AAQBmTgABAGZOAKFogwLUdmdzuM02yRNftzubClEBCqcRGcUb7H5pdzbT7mwmSRpLy2P3rDcfNyhrLC1/8pn1GT0hAEeliRLHMotGZPVU01xZ2g0WrzVTq9XSfW9z7J7aueV0u91s/bqZdmczzVZnJs3INCIDmIwlGE6F0RmPveFj78/2zo4AUB4BhKIN6jgGSy+51UmSLFTup7J4J5XFO1mo3B+7NrhXDQhAudSAUKzB0kuSNFudfOu73Xztoz/l82f7WajcGLt3p/9q/vV+Jf/o306ztTU8N8bhdABlsn7NsZxUDcjo4XO1zzzIF+u3c/brf3imz77/lx/kC2d/nJ3+i9lqt0/0cDo1IACT8Y8nx3ISAWT07JeFyv08/Hg9L33u5pHG6Pd+lkf/+/6JhxABBGAyakAoyt7wsfjCG0cOH0lSqV7O4gtvZKFy3wF1AAUSQCjG3lNvq2feTaV6eeLxKtXLqZ55d1qPB8AUmT7mWKa5BFNPdbidtrJ4J2c+/cOpjPvxf6+l/+jl1M4tT30ZxhIMwGTMgFCE0aLTJKmceWdqYw/GsgwDUA4BhGIMll6SpFJ9e2rjDsYaHR+A+RJAKM5C5f6+Ph/HG+/G42ZlABRBAGHu9i6/LFTuTf07TmJMACYngFCEWS2PqAMBKIMAQlHWm2v57W9+OfVxd/ovJVEHAlAKAYS5Giy/bLXbw5+99os/JsmtaX3Hw+43s9N/cVrDATAFAghF2DczsZPNaY390T+/Mq2hAJgSAYS5O3BZpJLfT2v8D299eVpDATAlAghFWjibzTePfgTMPm/eTP59+7PD9yfRDRWAoxNAmKvt9FI7t5xktwB11NXryZ0Pzk489r27F3L1usJTgBIJIBRjrbmextJymitLSbJz5Wpy/dqXcu/uhSOPde/uhVz8+c389HtrT78ZgJkTQCjK+UYjjaXl4fsP/r6Y7bdeSe/B6888Ru/B69l+65VcuTo++2H5BaAci/N+ANhr6eL68HWz1Umz1UmSPPzrtVTOvJNK9e19rdp3+q+m3/tO+h9/O/1HL6f7nw9n+swAHI0AQlEG/UC200s91WxsbGRlZSVJ8qlabXjfT36UfOOru6///LfkytUbScZDyWh7dwDKIoAwd7uFqNV039vcd20QPpKk2+0eOsavfrf7d6vVSpKsrq4qPgUomBoQijDYDbN0cf2pdRqDkHGQQWDZ2NgY7q5JMtZpFYD5E0AoxnZ6ExeJjoaS0VmTZDd8PEuwAWB2LMFw6qysrKTVag2DxmEzIqOzIMIHQFnMgHBq1Wq11Gq1rK6u7rs2uhQDQHnMgFC0Wq22r/i01WpldXV1bFajvlrd99mDggkAZRBAKFrn0lpqI9tvBw5aUtkbOD7pqDrsIwJAOQQQina+0Ujn0uP37c7moYFiEDgGBh1VBRCA8gggFG+0n8eTdrM0lpYP6f2xfsDPAJgnRaicGpP08tD/A6BMZkA4Fbba7bQ7m0+8Z3D9fKMxDB76fwCUaWHeD8Dptp3ezkmNXc/+9uyHnWhbz/5dMAMnGUDqqfodApiAGRCeGwf1/LAVF6BMAgjPhe30Ul+tjvUMqdVqll8ACiWAULTRA+WS3VmO+mr1wGCxnV5qterYewDKZBcMRaqnmo2NjXS73bE/TzM40E74ACibAEJxBuFj76m2ADw/BBCK8qTwcdAZMACcTrYQcizT3oZbT/XApZZSw4dtuACTUYRKUfYWku69BgAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAADxH/g92A+tBgH+59gAAAABJRU5ErkJggg=="`;

exports[`draw 2`] = `"data:image/png;base64,iVBORw0KGgoAAAANSUhEUgAAAiAAAAFgCAYAAACCOcLtAAAABmJLR0QA/wD/AP+gvaeTAAAFEElEQVR4nO3bPYudRRgG4Hv8iBhZAskuRCHE2kIIijaCYMDCKhAMKUQMEkGLNAHtxMbGQgRBU0RFbMSAmCpooz9AEZtoymS1EUHFmIgWt8W+kXg8fiSsexb2uuAUZ+aZ5505nOJh3pkEAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAACA69H21raXFj0PAP7qhkVPAADYehQgAMCGU4CwUF3zTNuzbX9o+27bHW0fb/t224Nt32h7W9uLM2P/aGu7u+2bbd9qe7rt8SlstD3e9tyU/522S9OYR9t+1/aXth+33bnBywfYshQgbAYrSR5JckeSs0lenNrvT/LVGOPJfxrcdiR5L8lrY4wjSQ4k2Z/koSTbkiwleTjJniTfJnk+ScYYp8YYy0l2Jbmc5Kn1XRYAsCm17cz3lbYXruyAXNX+tzsgbfe1/WSm7+m2r84eQm17Z9tzc+ZxrO3r67IoAP7VTYueAMz4McnyNY7Zm+TB2WImyftzYr+/kr/t0SSHk9yVtV2Yk9f4XACuk1cwbDa7kqzOaf8tyba28/6zq0nOjBlJHpsTu5LkQtv9SZ5N8lKSfUmOrc/0AfgvFCAsXNv72t7SdneSQ0k+mo0ZY/ya5MskT7Td3vbuJK9M3V8kWWr7wNS31PZwkueS3Nj23in/cpKjST5IsiPJN0k+z1rRc+j/XicAsElMt2A+bPtz26/bvjwVEX86AzLF3tP20yn2s7ZHrroFc/t0a2a17U/T+Y+dbS9P+S+2Pd/2hbY3T59TbS9NuU62PbGYXwEA2FBzzm0AsAV4BQMAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAAwNb2Ow1qeH93s2dWAAAAAElFTkSuQmCC"`;
